data/transactions/logic/TEAL_opcodes.md

```diff
@@ -1560,3 +1560,71 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
 | 0 | BlkSeed | []byte |  |
 | 1 | BlkTimestamp | uint64 |  |
 
+
+## ec_add g
+
+- Opcode: 0xe0 {uint8 curve}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
+- for curve points A and B, return the curve point A + B
+- **Cost**:  BN254g1=310 BN254g2=430 BLS12_381g1=540 BLS12_381g2=750
+- Availability: v9
+
+`EC` Groups:
+
+| Index | Name | Notes |
+| - | ------ | --------- |
+| 0 | BN254g1 | G1 of the BN254 curve. Points encoded as 32 byte X following by 32 byte Y |
+| 1 | BN254g2 | G2 of the BN254 curve. Points encoded as 64 byte X following by 64 byte Y |
+| 2 | BLS12_381g1 | G1 of the BLS 12-381 curve. Points encoded as 48 byte X following by 48 byte Y |
+| 3 | BLS12_381g2 | G2 of the BLS 12-381 curve. Points encoded as 96 byte X following by 48 byte Y |
+
+
+A and B are curve points in affine representation: X concatenated with Y. Fails if A or B is not in G. Does _not_ check if A and B are in the main prime-order subgroup.
+
+## ec_scalar_mul g
+
+- Opcode: 0xe1 {uint8 curve}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
+- for curve point A and scalar B, return the curve point BA
+- **Cost**:  BN254g1=2200 BN254g2=4460 BLS12_381g1=3640 BLS12_381g2=8530
+- Availability: v9
+
+A is a curve point encoded and checked as described in `ec_add`. Scalar B is interpreted as a big-endian unsigned integer. Fails if B exceeds 32 bytes.
+
+## ec_pairing_check g
+
+- Opcode: 0xe2 {uint8 curve}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., uint64
+- 1 if all points in A (concatenated points in group G) are paired with the corresponding points of B (concatenated points in the associated group), else 0
+- **Cost**:  BN254g1=18000 BN254g2=18000 BLS12_381g1=15000 BLS12_381g2=15000
+- Availability: v9
+
+A and B are concatenated points, encoded and checked as described in `ec_add`. A contains points of the group G, B contains points of the associated group (G2 if G is G1, and vice versa). Fails if A and B have a different number of points, or if any point is not in its described group or outside the main prime-order subgroup - a stronger condition than other opcodes.
+
+## ec_multi_exp g
+
+- Opcode: 0xe3 {uint8 curve}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
+- for curve points A and scalars B, return curve point B0A0 + B1A1 + B2A2 + ... + BnAn
+- **Cost**:  BN254g1=800 BN254g2=1800 BLS12_381g1=1400 BLS12_381g2=3500
+- Availability: v9
+
+A is a list of concatenated points, encoded and checked as described in `ec_add`. B is a list of concatenated 32 bytes scalars.
+
```

Comment:
```suggestion
Contrary to `ec_scalar_mul`, all scalars must be exactly 32-bytes in `ec_multi_exp`.

Note: the name `ec_multi_exp` comes from https://eips.ethereum.org/EIPS/eip-2537 and GNARK, but a more consistent name would be `ec_multi_scalar_mul`.
```

Reply:
I'd rather avoid explicitly referencing the EIP and GNARK directly.  How about:
The name `ec_multi_exp` was chosen to reflect common usage, but a more consistent name would be `ec_multi_scalar_mul`?

Separately. Should we just get rid of `ec_scalar_mul`? You should get the exact same answer using this opcode with one point and one argument. (Internally, surprisingly, we might still call a different gnark routine for the singleton case because it's faster.). This also gets rid of the discussion of the different scalar sizes.

It we do, could we call it `ec_dot_product` and also avoid this discussion?

Reply:
I up the proposal of renaming to `ec_dot_prod` than `ec_multi_exp`, and probably merge this one against `ec_scalar_mult`, for that is just one entry in vector multiplying another entry in another vec.

We might want to have linear cost model for this one? For the computation varies by the byte length.

Reply:
Currently proposing `ec_multi_scalar_mul` is lower conversation, resolving this one.